verification/src/sapling.rs

```diff
@@ -0,0 +1,302 @@
+		return Err(OutputError::Proof(ProofError::Failed));
+	}
+
+	Ok(())
+}
+
+fn accept_sapling_final(sighash: &[u8; 32], total: Point, sapling: &Sapling) -> Result<(), Error> {
+	// obtain current bvk from the context
+	let mut binding_verification_key = redjubjub::PublicKey(total);
+
+	// compute value balance
+	let mut value_balance = compute_value_balance(sapling.balancing_value)?;
+
+	// subtract value_balance from current bvk to get final bvk
+	value_balance = value_balance.negate();
+	binding_verification_key.0 = binding_verification_key.0.add(&value_balance, &JUBJUB);
+
+	// compute the signature's message for binding_verification_key/binding_sig
+	let mut data_to_be_signed = [0u8; 64];
+	binding_verification_key.0.write(&mut data_to_be_signed[..32]).expect("bvk is 32 bytes");
+	data_to_be_signed[32..].copy_from_slice(&sighash[..]);
+
+	// deserialize the binding signature
+	let binding_sig = Signature::read(&sapling.binding_sig[..])
+		.map_err(|err| Error::BindingSig(SignatureError::Invalid(err)))?;
+
+	// check the binding signature
+	let is_verification_ok = binding_verification_key
+		.verify(&data_to_be_signed, &binding_sig, FixedGenerators::ValueCommitmentRandomness, &JUBJUB);
+	if !is_verification_ok {
+		return Err(Error::BindingSig(SignatureError::Failed));
+	}
+
+	Ok(())
+}
+
+// This function computes `value` in the exponent of the value commitment base
+fn compute_value_balance(value: i64) -> Result<Point, Error> {
+	// Compute the absolute value (failing if -i64::MAX is the value)
+	let abs = match value.checked_abs() {
+		Some(a) => a as u64,
+		None => return Err(Error::InvalidBalanceValue),
+	};
+
+	// Is it negative? We'll have to negate later if so.
+	let is_negative = value.is_negative();
+
+	// Compute it in the exponent
+	let mut value_balance = JUBJUB
+		.generator(FixedGenerators::ValueCommitmentValue)
+		.mul(FsRepr::from(abs), &JUBJUB);
+
+	// Negate if necessary
+	if is_negative {
+		value_balance = value_balance.negate();
+	}
+
+	// Convert to unknown order point
+	Ok(value_balance.into())
+}
+
+/// Reads an FrRepr from a [u8] of length 32.
+/// This will panic (abort) if length provided is
+/// not correct.
+fn read_le(from: &[u8; 32]) -> FrRepr {
+	let mut repr = FrRepr::default();
+	repr.read_le(&from[..]).expect("length is 32 bytes");
+	repr
+}
+
+
+/// Deserializes point from the serialized buffer, checking that it is on the curve
+/// AND it MUST NOT be of small order.
+fn require_non_small_order_point(point_buff: &[u8; 32]) -> Result<Point, PointError> {
+	let point = Point::read(&point_buff[..], &JUBJUB).map_err(PointError::Invalid)?;
+	if is_small_order(&point) {
+		return Err(PointError::SmallOrder);
+	}
+	Ok(point)
+}
+
+/// Is this a small order point?
+fn is_small_order(point: &Point) -> bool {
+	point.double(&JUBJUB).double(&JUBJUB).double(&JUBJUB) == edwards::Point::zero()
+}
+
+#[cfg(test)]
+mod tests {
+	// TODO: detailed tests when sighash + verification keys are available
```

Comment:
isn't sighash already ok?

it is used in script signature verification and looks ok (after #4 )

Reply:
iiuc sighash format has been changed in overwinter && sapling upgrades: https://github.com/zcash/zips/blob/master/zip-0143.rst && https://github.com/zcash/zips/blob/master/zip-0243.rst